src/google/protobuf/compiler/csharp/csharp_message_field.cc

```diff
@@ -171,6 +171,14 @@ void MessageOneofFieldGenerator::GenerateMembers(io::Printer* printer) {
     "}\n");
 }
 
+void MessageOneofFieldGenerator::GenerateMergingCode(io::Printer* printer) {
```

Comment:
I think the original GenerateMergingCode method on line 19 is correct. This one has problems same with PrimitiveOneofFieldGenerator I mentioned bellow. I think the reason of the issue is somewhere else cleared the message field

Reply:
No, the original code checks `other` without any need - because again, with the change I've made, the oneof merge code is *only* called when `other` is already for the field in question.

The issue is because prior to this PR, the merge for a `oneof` field was just an assignment, with no merging. It's now a simple assignment in some cases, and a merge where appropriate.

Reply:
Oh you are right I missed your change of csharp_message.cc which is the real cause of the issue. For none oneof fields, can we change the check of other into csharp_meesge.cc too? Then oneof and normal fields can share the GernerateMergingCode  which is same with what we c++ code does.

Reply:
I'll have a look and see what I can do. I think I'll at least keep that as two separate commits - I'd like that step to be a pure refactoring that doesn't affect the generated code. (It's probably worth waiting until we've got #3798 in before this one, too, so that we can regenerate all the code at the same time.)

Reply:
Hmm. I think the C++ code is different already because the message fields aren't pointers, but I could be very wrong. (I have very little experience with C++ in general, and C++ protos in particular.)

I note that the Java code looks (a bit) like the C# code after this commit:

```java
void ImmutableMessageFieldGenerator::
GenerateMergingCode(io::Printer* printer) const {
  printer->Print(variables_,
    "if (other.has$capitalized_name$()) {\n"
    "  merge$capitalized_name$(other.get$capitalized_name$());\n"
    "}\n");
}
...
void ImmutableMessageOneofFieldGenerator::
GenerateMergingCode(io::Printer* printer) const {
  printer->Print(variables_,
    "merge$capitalized_name$(other.get$capitalized_name$());\n");
}
```

Reply:
(Ditto for primitive fields in Java.)

Reply:
LGTM as it is similar with Java